### ISSUE TITLE
Promtail keeps files locked 

### ISSUE THREAD

**Issue by zlepper** [grafana/loki#3668]
**Describe the bug**
See #2743 and #2209 both of which are closed because they are "stale", however they are very much valid bug reports, and are currently killing servers for us...

**To Reproduce**
1. Started Loki 2.2.1
2. Started Promtail 2.2.1 to tail out log directory recursively 'G:\LogFiles\Digizuite\dam.dev.digizuite.com\**\*.log'
3. Create a lot of logs
4. Attempt to delete said logs manually

**Expected behavior**
I expect to be able to delete the log files without Promtail holding a lock on them. 

**Environment:**
 - Infrastructure: Virtual machine, Windows server 2012 R2 tandard
 - Deployment tool: Manually configured windows service.

**Comment by mateinavidad** [comment]
I am seeing a similar behaviour, i did some digging in the code, and they use the [hpcloud/tail](github.com/hpcloud/tail) project for tailing files, which has not received any updates in the last 3 years and is considered dead. 

The project was revamped under [nxadm/tail](https://github.com/nxadm/tail) which is still maintained. 
I created a local build replacing the hpcloud implementation with nxadm, and the needed code adjustments, but the library suffers from the same locking issue. 

So the problem is the go library for emulating tail -f on windows.

**Comment by zlepper** [comment]
Ping.

**Comment by zlepper** [comment]
Ping.

**Comment by zlepper** [comment]
Ping

**Comment by zlepper** [comment]
Ping

**Comment by vittoriopelle** [comment]
On the same bug, need to read a file that have to be written but promtail block the file so my process can't write other log into it.
I'm on windows 10 20H2.

**Comment by naveenkumarsp** [comment]
I too have the same problem where promtail is locking the file and not allowed to rotate or delete

**Comment by perosb** [comment]
Similar here that pods gets stuck in "Terminating" status due to cannot remove .log file.

**Comment by perosb** [comment]
I did a simple test with:

```golang
package main
import (
	"fmt"
	/*"github.com/nxadm/tail"*/
	"github.com/hpcloud/tail"
)
func main() {
  t, err := tail.TailFile("kermit.log", tail.Config{Follow: true, Poll: true, ReOpen: true, MustExist: true, Location: &tail.SeekInfo{Offset: 0, Whence: 0}})
  if err != nil {
      panic(err)
  }
  for line := range t.Lines {
      fmt.Println(line.Text)
  }
}
```

I'm not able to reproduce the locking of the files with that.

**Comment by slim-bean** [comment]
Apologies for never leaving any specific comments on this issue, but that's largely because I don't really have great news here. The problem is how windows filesystems works and is an annoying issue.

On most linux filesystems when you open a file you get a "handle" to that file which still allows for operations like rename/delete to be performed on that file. The underlying file is kept available as long as their are processes which have handles to it. Thus you can delete a file or rename a file which is used by a process without issue on most (maybe all) nix filesystems.

Windows on the other hand is not this flexible, when a file is opened for reading it basically gets and exclusive lock which then blocks any other operations like deleting or renaming.

This would require either coordinating a file rotation with a restart of promtail, or I guess some kind of periodic close/re-open of files within promtail?

I'm not sure if anyone has more experience with windows and what other solutions are available but I don't think there is a simple fix to this as the issue stems from the filesystem behavior on Windows not being nearly as flexible as other filesystems...

**Comment by zlepper** [comment]
> Apologies for never leaving any specific comments on this issue, but that's largely because I don't really have great news here. The problem is in how windows filesystems work.
> 
> On most linux filesystems when you open a file you get a "handle" to that file which still allows for operations like rename/delete to be performed on that file. The underlying file is kept available as long as their are processes which have handles to it. Thus you can delete a file or rename a file which is used by a process without issue on most (maybe all) nix filesystems.
> 
> Windows on the other hand is not this flexible, when a file is opened for reading it basically gets and exclusive lock which then blocks any other operations like deleting or renaming.
> 
> "Fixing" this would require either coordinating a file rotation with a restart of promtail, or I guess some kind of periodic close/re-open of files within promtail?
> 
> I'm not sure if anyone has more experience with windows and what other solutions are available but I don't think there is a simple fix to this as the issue stems from the filesystem behavior on Windows not being nearly as flexible as other filesystems...

That is not entirely correct. When you open a file on Windows you can specify the sharing mode, aka what other processes and such should be able to do, and one of these modes does allow for deletion, without breaking the current process.

**Comment by slim-bean** [comment]
>That is not entirely correct. When you open a file on Windows you can specify the sharing mode, aka what other processes and such should be able to do, and one of these modes does allow for deletion, without breaking the current process.

Ooooh, interesting, I wonder now how/if we can do this from Go

**Comment by MasslessParticle** [comment]
This has historically been a known issue with golang and we've verified that our tailing library implements all the necessary workarounds to not lock files.

I've tested across versions of promtail and golang

I've attempted to reproduce this on a windows machine without luck in the following ways:
- Tail the whole directory
- Tail single files

I got a locking error when I tailed a directory and continuously wrote to a log file in it. When I tried to delete the file, Windows complained that promtail held a lock even though another process was holding it open. I don't think this is our exact issue, however, because the file was never in the `DeletePending` state.

I am unable to get a file stuck in the `DeletePending` state

I think we have an experiment we can do where promtail checks for a DeletePending when tailing but we'll need help to test it. @zlepper @perosb any chance you or anyone else may be able to test?

**Comment by zlepper** [comment]
@MasslessParticle I'm pretty sure I can still reproduce the issue on one of our internal servers, so yes, I would definitely be up to help testing.

**Comment by slim-bean** [comment]
@zlepper quick question, how does your file rotation work on the affected servers? Is it the process that generates the files that does the rotation or is there an external process which does it?

**Comment by zlepper** [comment]
@slim-bean it's the same process. Specifically it creates the file, writrs some stuff, at some point discovers that the file is over a configured threshold, at which point it closes the file, and moves on to a new file. After it has moved to the new file it does a file listing in the active log directory, and then attempts to delete older log files until the file count is below a configured limit.

**Comment by MasslessParticle** [comment]
@zlepper I've got an experimental branch with a change to tail that *should* detect a file in `DeletePending` state and close it. I've tested it on my Windows machine to verify that it doesn't break anything but can't reproduce the issue so I don't know if it actually fixes anything.

https://github.com/grafana/loki/compare/main...tpatterson/promtail-windows-experiment

Are you setup to build go projects?

**Comment by zlepper** [comment]
> @zlepper I've got an experimental branch with a change to tail that *should* detect a file in `DeletePending` state and close it. I've tested it on my Windows machine to verify that it doesn't break anything but can't reproduce the issue so I don't know if it actually fixes anything.
> 
> https://github.com/grafana/loki/compare/main...tpatterson/promtail-windows-experiment
> 
> Are you setup to build go projects?

Yes, I can setup and build locally, will try your branch and see what happens.

**Comment by zlepper** [comment]
@MasslessParticle 
Of cause I can't reproduce the original locking issue at all right now :/ Just wanted to verify before I tested the branch you provided. I will keep trying during the week, just so you know i haven't forgotten.

**Comment by MasslessParticle** [comment]
@zlepper Thank you!

**Comment by tiger5226** [comment]
@zlepper / @MasslessParticle  We are currently experiencing this problem in production causing logs to backup on servers. We can also deploy the build and test it out as well. It is a continuous problem, so it should reproduce pretty quickly.

**Comment by MasslessParticle** [comment]
@tiger5226 Any testing would be much appreciated!

**Comment by tiger5226** [comment]
Hi Team, 

The issue persists with the branch @MasslessParticle 

We have had it running in production for a few days now. This morning we noticed that the folder size of the log files was increasing from the starting point. So we tested a restart and sure enough upon restart the log files were removed. So it is still holding on to them with the changes. 

As a temporary mitigation we are going to see if we can run in the task scheduler and basically restart it daily, to see if this reduces the build up of log files. The chief concern here is that the files build up so much that the server runs out of disk space. This happened to us in production already, which is why I found this issue thread. 

Any other suggestions? Happy to help out here, like if you come up with a debug build etc.

**Comment by MasslessParticle** [comment]
@tiger5226 Interesting, I wonder what's unique about your environment. Can you check if the file is in a `DeletePending` state of the file when it's stuck?

**Comment by tiger5226** [comment]
ok, before putting the auto restart in, we will check for DeletePending state.

**Comment by tiger5226** [comment]
@MasslessParticle We let it run for a couple of days and checked with procmon the state of the files. They are indeed in DeletePending state. When we restart the binary they get deleted.

**Comment by tiger5226** [comment]
Another interesting note, I do believe environment may have an impact here. Even without this PR, there are machines that do not have this problem. It is not universal. Before you ask, we tried to find a commonality but the machines are nearly identical and software producing the log files is the same as well. A bit of needle in a haystack trying to find a meaningful difference. If you have any clues around this specific issue, I am happy to investigate further.

**Comment by solo-untld** [comment]
I have same problem.. BUT everything was normal. Script was able to write log anf promtail read it. No access troubles. After one month of down I start promtail again and got this error... Between that i've install updates on server Windows Server 2019. I'm not able to uninstall updates from this server, cause it's critical production server. Maybe Microsoft makes any update it security politics? maybe it's right way to see updates difference in your test systems?

**Comment by solo-untld** [comment]
I've got some update for that. PowerShell scripts still able to write log without any errors, but only using 7 version. 5.1 shell has same problem

**Comment by JStickler** [comment]
Promtail is end of life (EOL) as of  March 2, 2026.  No future support or updates will be provided. All future feature development will occur in [Grafana Alloy](https://github.com/grafana/alloy).
